### ISSUE TITLE
CancelledError on LTA retrieval with download_all

### ISSUE THREAD

**Issue by rbavery** [sentinelsat/sentinelsat#425]
I'm getting CancelledError when trying this code [from the docs](https://sentinelsat.readthedocs.io/en/latest/api_overview.html) with my own aoi which is a valid geojson that I confirmed by plotting in QGIS:
```python
from sentinelsat import SentinelAPI, read_geojson, geojson_to_wkt
from datetime import date

# enter credentials in .netrc https://sentinelsat.readthedocs.io/en/master/api_overview.html
api = SentinelAPI("user", "pw", 'https://scihub.copernicus.eu/dhus')

# search by polygon, time, and SciHub query keywords
footprint = geojson_to_wkt(read_geojson("data/santa_cruz_bolivia_forest_site_WGS84.geojson"))

products = api.query(footprint,
                     date=('20141001', '20141130'),
                     platformname='Sentinel-1',
                    producttype='GRD')

# download all results from the search
api.download_all(products_df.index.to_list())
```

The docs say this:

```python
Signature:
api.download_all(
    products,
    directory_path='.',
    max_attempts=10,
    checksum=True,
    n_concurrent_dl=2,
    lta_retry_delay=600,
)
Docstring:
Download a list of products.

Takes a list of product IDs as input. This means that the return value of query() can be
passed directly to this method.

```

But `products` is an `OrderedDict`

and a list of ids also results in a Cancelled Error. I'm not sure what sthese should be so I tried the `title`, `index`, and `link` from the OrderedDict converted to a pandas dataframe.

```python
# download all results from the search
products_df = api.to_dataframe(products)
api.download_all(products_df.index.to_list())
```

This is the error

```python
---------------------------------------------------------------------------
CancelledError                            Traceback (most recent call last)
<ipython-input-71-87e3b62cde07> in <module>
      1 # download all results from the search
----> 2 api.download_all(products_df.index.to_list())

/usr/local/lib/python3.6/dist-packages/sentinelsat/sentinel.py in download_all(self, products, directory_path, max_attempts, checksum, n_concurrent_dl, lta_retry_delay)
    777 
    778             for dl_task in concurrent.futures.as_completed(dl_tasks):
--> 779                 if not dl_task.exception() and dl_task.result():
    780                     product_info = dl_task.result()
    781                     downloaded_prods[product_info["id"]] = product_info

/usr/lib/python3.6/concurrent/futures/_base.py in exception(self, timeout)
    454         with self._condition:
    455             if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:
--> 456                 raise CancelledError()
    457             elif self._state == FINISHED:
    458                 return self._exception

CancelledError: 
```

**Comment by j08lue** [comment]
Your usage is correct. When you pass in the complete `products` dictionary, `download_all` will only use the keys from the dictionary, which are the IDs:

https://github.com/sentinelsat/sentinelsat/blob/9c6269c9ec04e1d2f41ac671d75caf23077bbc25/sentinelsat/sentinel.py#L733

The only thing you might want to change is to run against `/dhus` rather than the recommended `/apihub`.

Regarding the error you get: I have no clue whatsoever why your download job future (`concurrent.futures` asynchronous task) gets cancelled before completing or failing - which is giving you the `CancelledError`.

If you are unable to resolve this somehow, a workaround you might want to use is to just download the products sequentially by looping over them and calling `api.download` on each of them. I guess you are using two concurrent threads anyways, so it will only take about twice as long. Like (with another progress bar):

```python
import tqdm

for pid in tqdm.tqdm(products):
    api.download(pid, ...)
```

**Comment by heyarne** [comment]
I am facing the same error with Sentinel-2 L2A data. The API currently seems a little unstable to me, given the other issue I described in #427. After querying the API I manipulating the resulting products in a geodataframe (`gdf`). I'm then trying to download the products like so:

``` python
api.download_all(gdf['uuid'].values)
```

The UUIDs are as follows:
<details>
<summary>List of UUIDs</summary>

```
3318d298-429c-419d-a141-ec4c2f1820db
2b3f3844-5865-4954-b300-575aa09459d5
b9903874-93ef-4520-bf49-951e810484f3
b92db34e-877e-447c-9ff7-078cd0eb7208
57b8546e-ab1f-4916-89df-a2b63322f358
ef822323-fe13-4657-be94-f8b5c519af89
6f9119c9-88b6-44b1-8018-32ca0d783ced
3ad8da72-dc77-4c79-b7e2-4acc2e627725
0fb1f9ad-c7ac-4536-8ada-a135717cce9e
708574e6-3bea-46be-a5e0-2610f61d6571
162c5278-9b04-48e1-b0de-5130a7d76310
326d46e3-d83c-422b-9c30-0601a93d4278
a686736d-5152-40e2-9cf5-ddf2c2fafd8e
b32a4b94-9ff4-4929-9731-2b63a068b456
c63e6240-15c6-49c5-bfef-e77072f1e2d0
4686525f-ae56-4494-94db-77c84f4e71f2
d64991c3-5d9a-4616-9340-57d3461211f1
69035a27-e2f3-43a7-a669-5cc10549fa1f
9ffe718e-ec45-4a07-88ae-7d8371c94073
0347194e-71bd-4b8d-aee3-217a25d84b42
bedec483-5ee1-4264-8dfa-a3b53ce364f7
67780dcc-1829-4d6d-9cca-c58347727024
391acf3e-b471-4db8-a1b8-4d43a318f8f0
c51bbb1b-a5b0-477d-9948-11d8b6689e4e
6c9cbc5d-d524-4705-b548-cc0641f45bfa
bce0a851-6e98-4632-9c76-02fcb283361e
c875eccb-f8ce-4a01-b112-8bcad8d97386
6d165df0-5775-4856-8e39-a7112a0be142
2f49ee17-4cb1-43cf-8387-555ad282eb3c
eb56840a-d0b6-4fc9-88f9-40984d44446d
7dcc1be8-3cad-4064-b80a-ddc57dc536eb
70540b28-d8da-439d-a20f-d44fc415c7fa
9677ffca-cdff-4461-879f-e673b20eda4f
b7bf6a53-08d8-4de8-b47f-0f3edebc2e81
5805bcbc-04d9-4c46-a26d-6471655cba13
b1957288-84b8-4257-8a53-42ee90a34f9e
2901f259-9bd1-4256-b45b-a011ee36a9ab
338fb5ea-7887-40ff-b805-91f5cd25e267
335edaf8-81ef-4d44-8088-e1cc40a3a835
4a2ec86c-9a81-46db-a3a6-69fb81f4622a
```
</details>

Note that some of these products have to be retrieved from the LTA. I received the LTA notification when trying to download the products in the loop you suggested as a workaround.

**Comment by j08lue** [comment]
Please let the Copernicus team know, also by responding to this survey (by 15 March): https://sentinels.copernicus.eu/web/sentinel/-/copernicus-sentinel-data-access-and-user-services-user-satisfaction-survey-2020

**Comment by ervingrebesh** [comment]
@j08lue  does I understood correctly, that some time ago download_all worked fine with getting data from LTA? 
Or you don't have any problems even now calling download_all? 
Just would like to know where is probable error: user level/ code level / copernikus ?

Looping api.download won't work perfectly because it won't find already online products and there is no mechanism to wait in queue.  If you are trying to download different LTAs to often you will be dropped from connection.

**Comment by mcgregorian1** [comment]
I can confirm running into the same issue as mentioned above for Sentinel 2 images. @j08lue is there not a workaround for this until Copernicus does something on their end?

**Comment by j08lue** [comment]
This is really not a trivial error (race condition). A lead to debugging this might come from a thread from Dask: https://github.com/dask/distributed/issues/4612#issuecomment-604417345 There, the issue was that futures were deleted and re-created with the same parameters in very short succession. Could this also be the case here?

**Comment by j08lue** [comment]
The rapid resubmission of identical futures might originate from this retry loop:
https://github.com/sentinelsat/sentinelsat/blob/9c6269c9ec04e1d2f41ac671d75caf23077bbc25/sentinelsat/sentinel.py#L842-L859

But I must admit that I have never gotten fully into the logic behind the dual thread pool for triggering and downloading. Maybe the author @gbaier could comment here? (original PR was #297)

**Comment by valgur** [comment]
Should be fixed with #448 and #479 by @avalentino.
